include/namedstruct/phantom_type.h

```diff
@@ -0,0 +1,238 @@
+#include <utility>
+#include <optional>
+
+#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
+#include <msgpack.hpp>
+#endif
+
+namespace namedstruct {
+/*
+ * A strong type alias for a numerical value (e.g. an ID or quantity). Can only be
+ * constructed explcitly. Arthimetic operations and conversions back to T are implicit.
+*/
+template <typename T, typename Derived> struct phantom_type {
+    using value_type = T;
+
+    phantom_type() = default;
+    constexpr explicit phantom_type(const T& value) : value(value) {}
+    constexpr explicit phantom_type(T&& value) : value(std::move(value)) {}
+
+    T value;
+
+    // Implicitly convert to the underlying value
+    operator T&() noexcept {
+        return value;
+    }
+    constexpr operator const T&() const noexcept {
+        return value;
+    }
+
+    // Operators
+    T operator++() {
+        return ++value;
+    }
+    T operator++(int) {
+        return value++;
+    }
+
+    T operator--() {
+        return --value;
+    }
+    T operator--(int) {
+        return value--;
+    }
+
+    // Operators using derived type
+    constexpr auto operator+(const Derived& rhs) const {
+        return Derived(value + rhs.value);
+    }
+
+    constexpr auto operator-(const Derived& rhs) const {
+        return Derived(value - rhs.value);
+    }
+
+    constexpr auto operator*(const Derived& rhs) const {
+        return Derived(value * rhs.value);
+    }
+
+    constexpr auto operator/(const Derived& rhs) const {
+        return Derived(value / rhs.value);
+    }
+
+    constexpr auto operator<(const Derived& rhs) const {
+        return value < rhs.value;
+    }
+
+    constexpr auto operator==(const Derived& rhs) const {
+        return value == rhs.value;
+    }
+
+    auto& operator+=(const Derived& rhs) {
+        value += rhs.value;
+        return static_cast<Derived&>(*this);
+    }
+
+    auto& operator-=(const Derived& rhs) {
+        value -= rhs.value;
+        return static_cast<Derived&>(*this);
+    }
+
+    auto& operator*=(const Derived& rhs) {
+        value *= rhs.value;
+        return static_cast<Derived&>(*this);
+    }
+
+    auto& operator/=(const Derived& rhs) {
+        value /= rhs.value;
+        return static_cast<Derived&>(*this);
+    }
+
+    // limits
+    static constexpr Derived max() noexcept {
+        return Derived(std::numeric_limits<T>::max());
+    }
+
+    static constexpr Derived min() noexcept {
+        return Derived(std::numeric_limits<T>::min());
+    }
+
+    template <typename Y = T>
+    static constexpr Derived infinity(typename std::enable_if<std::is_floating_point<Y>::value>::type* = nullptr) {
+        return Derived(std::numeric_limits<Y>::infinity());
+    }
+
+    template <typename Other>
+    inline constexpr auto as() const {
+        return static_cast<Other>(value);
+    }
+
+#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
+    void msgpack_pack(msgpack::packer<msgpack::sbuffer>& pk) const;
+
+    void msgpack_unpack(const msgpack::object& o);
```

Comment:
Sorry, I wrongly said it was ok to merely declare the method stubs, but this won't work if the method bodies aren't defined:
```suggestion
    void msgpack_pack(msgpack::packer<msgpack::sbuffer>& pk) const {
        pk.pack(value);
    }

    void msgpack_unpack(const msgpack::object& o) {
        value = o.as<value_type>();
    }
```